query.go

```diff
@@ -174,7 +174,24 @@ func (dht *IpfsDHT) runQuery(ctx context.Context, target string, queryFn queryFn
 }
 
 func (q *query) recordPeerIsValuable(p peer.ID) {
-	q.dht.routingTable.UpdateLastUsefulAt(p, time.Now())
+	if !q.dht.routingTable.UpdateLastUsefulAt(p, time.Now()) {
+		// not in routing table
+		return
+	}
+
+	// Protect useful peers, when they're actually useful. This will last
+	// through disconnects. However, we'll still evict them if they keep
+	// disconnecting from us.
+	//
+	// Restrict to buckets 0, 1 (75% of requests, max 40 peers), so we don't
+	// protect _too_ many peers.
+	commonPrefixLen := kb.CommonPrefixLen(q.dht.selfKey, kb.ConvertPeerID(p))
+	cmgr := q.dht.host.ConnManager()
+	if commonPrefixLen < 2 {
+		cmgr.Protect(p, dhtUsefulTag)
+	} else {
+		cmgr.TagPeer(p, dhtUsefulTag, usefulConnMgrScore)
+	}
```

Comment:
@Stebalien The `TagPeer` call is idempotent and overwrites the existing score for that tag.
How about adding to the current score here so we "reward" peers who useful to us often.

Reply:
While we could probably do that if we utilize decaying tags like pubsub is planning on doing (https://github.com/libp2p/go-libp2p-pubsub/issues/306). However, I don't think it's currently worth the complexity especially if we take into account the various ways the connection manager can be gamed to promote one peer over another.

Reply:
I don't want to give them _too_ much weight, I just want to make sure that they're marked useful.

Reply:
it requires advanced knowledge of the code and its history along with gathering the correct key responses in order to game the authentication at this stage. the trust can be weighed by all peers with the already trusted members/organizations given the only authority to bring members into the core of the repository's they have trust with. this is also effective with machine learning agents/ai bots